ast.ts

```diff
@@ -15,39 +20,57 @@ export type Class<A> = { a?: A, name: string, fields: Array<VarInit<A>>, methods
 
 export type VarInit<A> = { a?: A, name: string, type: Type, value: Literal }
 
-export type FunDef<A> = { a?: A, name: string, parameters: Array<Parameter<A>>, ret: Type, inits: Array<VarInit<A>>, body: Array<Stmt<A>> }
+export type FunDef<A> = { 
+  a?: A, 
+  name: string, 
+  parameters: Array<Parameter<A>>, 
+  ret: Type, 
+  decls: Array<Scope<A>>,
+  inits: Array<VarInit<A>>, 
+  funs: Array<FunDef<A>>
+  body: Array<Stmt<A>>
+}
+
+export type Closure<A> = { 
+  a?: A, 
+  name: string, 
+  fields: Array<VarInit<A>>, 
+  apply: FunDef<A> 
+}
 
 export type Stmt<A> =
-  | {  a?: A, tag: "assign", name: string, value: Expr<A> }
-  | {  a?: A, tag: "return", value: Expr<A> }
-  | {  a?: A, tag: "expr", expr: Expr<A> }
-  | {  a?: A, tag: "if", cond: Expr<A>, thn: Array<Stmt<A>>, els: Array<Stmt<A>> }
-  | {  a?: A, tag: "while", cond: Expr<A>, body: Array<Stmt<A>> }
-  | {  a?: A, tag: "pass" }
-  | {  a?: A, tag: "field-assign", obj: Expr<A>, field: string, value: Expr<A> }
+  | { a?: A, tag: "assign", name: string, value: Expr<A> }
```

Comment:
Let's not have big whitespace changes in this PR.

Reply:
can you release the formatter? So that we can format before PR. Thanks.

Reply:
The correct order is to format **after** these PRs merge. I don't want a bunch of whitespace changes in everyone's PRs, I want everyone to re-pull correctly-formatted code. This changeset should not have whitespace changes.

The setup for the formatter is in https://github.com/ucsd-cse231-w21/chocopy-wasm-compiler/pull/3